File: backend/routes/interactions.py
from fastapi import APIRouter, Depends, status
from pydantic import BaseModel
from uuid import uuid4
from datetime import datetime, timezone
from sqlalchemy.ext.asyncio import AsyncSession

from database.db import get_db
from routes.auth import get_current_user
from database.models_sql import User

router = APIRouter(prefix="/interactions", tags=["interactions"])

class InteractionRequest(BaseModel):
    item_id: str
    interaction_type: str
    rating: int
    review_title: str
    review_content: str
    quantity: int

@router.post("", status_code=status.HTTP_201_CREATED)
async def log_interaction(
    interaction: InteractionRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    interaction_id = str(uuid4())
    timestamp = datetime.now(timezone.utc)

    # Simulated log (to be replaced with Kafka in the future)
    print({
        "interaction_id": interaction_id,
        "user_id": user.user_id,
        "item_id": interaction.item_id,
        "interaction_type": interaction.interaction_type,
        "rating": interaction.rating,
        "review_title": interaction.review_title,
        "review_content": interaction.review_content,
        "quantity": interaction.quantity,
        "timestamp": timestamp.isoformat()
    })

    return {
        "message": "Interaction logged (simulated)",
        "interaction_id": interaction_id
    }

File: frontend/src/services/interactions.ts
import { apiRequest, ServiceLogger } from './api';

export interface InteractionRequest {
  item_id: string;
  interaction_type: string;
  rating?: number;
  review_title?: string;
  review_content?: string;
  quantity?: number;
}

export interface InteractionResponse {
  message: string;
  interaction_id: string;
}

export const logInteraction = async (
  interaction: InteractionRequest
): Promise<InteractionResponse> => {
  ServiceLogger.logServiceCall('logInteraction', { interaction });
  return apiRequest<InteractionResponse>(
    '/api/interactions',
    'logInteraction',
    {
      method: 'POST',
      body: interaction,
    }
  );
};

export const recordProductClick = async (productId: string): Promise<void> => {
  ServiceLogger.logServiceCall('recordProductClick', { productId });
  return apiRequest<void>(
    `/api/products/${productId}/interactions/click`,
    'recordProductClick',
    {
      method: 'POST',
    }
  );
};
